### DISCUSSION TITLE
XRay trace ID conversion

### DISCUSSION THREAD

**Discussion by errordeveloper** [open-telemetry/opentelemetry-collector-contrib#5257] [Q&A]
I am looking at XRay exporter code, and I'm not quite sure how it's actually meant to work.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ac26625c2eea420d6703a15e9bd81d686fb5b8af/exporter/awsxrayexporter/internal/translator/segment.go#L257-L304

So XRay IDs are supposed to be much longer (35 bytes, not 16), however the code seems to ignore the length and just checks the epoch bytes... I'm not sure if this well intended, or a bug?

**Comment by Aneurysm9** [comment]
I believe this is correct.  Skipping the timestamp validation and just looking at the `traceID` construction:

```go
	content[0] = '1'
	content[1] = '-'
	hex.Encode(content[2:10], b[0:4])
	content[10] = '-'
	hex.Encode(content[identifierOffset:], traceIDBytes[4:16]) // overwrite with identifier

	return string(content[0:traceIDLength]), nil
```

`traceIDLength` is a constant `35`.  `content` starts as `[traceIDLength]byte{}`.

* 2 bytes for `1-`
* 8 bytes for a hex-encoded representation of the 4 byte timestamp
* 1 byte for `-`
* 24 bytes for a hex-encoded representation of the 12 bytes in the `traceID` following the timestamp

This fills all 35 bytes of `content` and then returns it as a string.

**Reply by errordeveloper** [reply]
In my case I am just ingesting spans with generated IDs (similarly to what [kspan](https://github.com/weaveworks-experiments/kspan/blob/12eefbe61283d2d6b0360fc39fc265076376c2d8/controllers/events/object.go#L109-L116) does). XRay exporter seems to reject these spans, I was hoping it would try convert the IDs by prepending a timestamp... what am I missing?

**Reply by Aneurysm9** [reply]
Unfortunately the `traceID` must be generated with the first 32 bits representing a timestamp if the spans are to be sent to XRay.  It is not feasible later to convert a random `traceID` to one that contains a timestamp as different spans from the same trace may end up being processed differently and thus with different timestamps injected into their `traceID`, which would break the trace.

`kspan` complicates things further in that it does not use randomly generated `traceID`s, but instead [constructs a `traceID` from a hash of object data](https://github.com/weaveworks-experiments/kspan/blob/12eefbe61283d2d6b0360fc39fc265076376c2d8/controllers/events/object.go#L109).  It may be possible to find a way to use a consistent timestamp there, but it's not immediately obvious.

**Reply by errordeveloper** [reply]
@Aneurysm9 thanks for explaining this. It makes sense to me.

I think that the documentation needs to be clearer about what exactly happens. Right now it's not very clear from the [readme](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ac26625c2eea420d6703a15e9bd81d686fb5b8af/exporter/awsxrayexporter/README.md#data-conversion). When I looked at the code linked above, the function name ([`convertToAmazonTraceID`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ac26625c2eea420d6703a15e9bd81d686fb5b8af/exporter/awsxrayexporter/internal/translator/segment.go#L257-L304)) implied conversion does take place and the documentation didn't make it very clear. Also, even the function argument is `pdata.TraceID`, so it must be attempting conversion, and that is actually the case as you described, however it's not sufficiently explicit that conversion will be done for trace IDs that do contain at timestamp already.

**Reply by Aneurysm9** [reply]
It looks like it was originally creating new `traceID`s with current timestamps, but that was changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1366 because it wasn't doing the right thing.  Unfotunately, the README wasn't also updated.

@mxiamxia are you, or someone from your team, able to review and update the documentation around this?
